pkg/cmd/cli/cmd/importimage.go

```diff
@@ -67,45 +68,147 @@ func RunImportImage(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, arg
 		return err
 	}
 
+	insecure := cmdutil.GetFlagBool(cmd, "insecure")
 	from := cmdutil.GetFlagString(cmd, "from")
 	confirm := cmdutil.GetFlagBool(cmd, "confirm")
-	insecure := cmdutil.GetFlagBool(cmd, "insecure-repository")
+	all := cmdutil.GetFlagBool(cmd, "all")
+
+	if len(from) == 0 {
+		from = streamName
+	}
+
+	ref, err := imageapi.ParseDockerImageReference(streamName)
+	if err != nil {
+		return fmt.Errorf("the image name must be a valid Docker image pull spec or reference to an image stream (e.g. myregistry/myteam/image:tag)")
+	}
+	if len(ref.ID) > 0 {
+		return fmt.Errorf("if you want to import an image by ID, use --from=%s %s", from, ref.AsRepository().Exact())
+	}
+	// apply the default tag
+	if !all && len(ref.Tag) == 0 {
+		ref.Tag = imageapi.DefaultImageTag
+	}
+	name := ref.Name
+
 	imageStreamClient := osClient.ImageStreams(namespace)
-	stream, err := imageStreamClient.Get(streamName)
+	stream, err := imageStreamClient.Get(name)
 	if err != nil {
-		if len(from) == 0 || !errors.IsNotFound(err) {
+		if !errors.IsNotFound(err) {
 			return err
 		}
-		if !confirm {
-			return fmt.Errorf("the image stream does not exist, pass --confirm to create")
+		if len(from) == 0 && !confirm {
+			return fmt.Errorf("no image stream named %q exists, pass --confirm to create and import", name)
 		}
-		stream = &imageapi.ImageStream{
-			ObjectMeta: kapi.ObjectMeta{Name: streamName},
-			Spec:       imageapi.ImageStreamSpec{DockerImageRepository: from},
+
+		if len(ref.Tag) == 0 && all {
+			stream = &imageapi.ImageStream{
+				ObjectMeta: kapi.ObjectMeta{Name: name},
+				Spec:       imageapi.ImageStreamSpec{DockerImageRepository: from},
+			}
+		} else {
+			stream = &imageapi.ImageStream{
+				ObjectMeta: kapi.ObjectMeta{Name: name},
+				Spec: imageapi.ImageStreamSpec{
+					Tags: map[string]imageapi.TagReference{
+						ref.Tag: {
+							From: &kapi.ObjectReference{
+								Kind: "DockerImage",
+								Name: from,
+							},
+						},
+					},
+				},
+			}
 		}
+
 	} else {
 		if len(stream.Spec.DockerImageRepository) == 0 && len(stream.Spec.Tags) == 0 {
 			return fmt.Errorf("image stream has not defined anything to import")
 		}
-		if len(from) != 0 {
-			if from != stream.Spec.DockerImageRepository {
+		if len(ref.Tag) == 0 && all {
+			if len(from) != 0 && from != stream.Spec.DockerImageRepository {
 				if !confirm {
+					if len(stream.Spec.DockerImageRepository) == 0 {
+						return fmt.Errorf("the image stream does not currently import an entire Docker repository, pass --confirm to update")
+					}
 					return fmt.Errorf("the image stream has a different import spec %q, pass --confirm to update", stream.Spec.DockerImageRepository)
 				}
 				stream.Spec.DockerImageRepository = from
 			}
+		} else {
+			var tag imageapi.TagReference
+			if existing, ok := stream.Spec.Tags[ref.Tag]; ok {
+				tag = existing
```

Comment:
The gap here was that from shouldn't be defaulted until we get into this chunk of logic.  If you try to change a tag by specifying a new --from, we should error out and tell you to use `oc tag` because you're trying to tag.  Cleaned up the structure and added more comments.